### PR TITLE
Add Cody verify your email popup and notice for dotcom

### PR DESCRIPTION
**Pull request by thenamankumar**
Add alerts, notices, popups dotcom for restricting Cody access to logged-in users with verified emails only. 

https://www.loom.com/share/865d8621dbd440a3ae926a647ce4adca

design: https://www.figma.com/file/DIk3eGdgKszigQVec7Y8SD/Frictionless-cody?type=design&node-id=767-30229&t=ee1VyzmpjtZhgrNq-0

## Test plan

- sign up with a new account

### CODE REVIEW

**Comment by sourcegraph-buildkite** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.24% (+6.68 kb) 🔺 | 0.02% (+3.08 kb) | -0.03% (-3.60 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/naman/cody-verify-email/compare-report.html;tab=live_object) for a full comparison between the commits 1790322 and 246cf14 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Review by rrhyne — APPROVED** [review]
Looks great Naman!

**Review by abeatrix — APPROVED** [review]

File: client/web/src/cody/stores/chat.ts [hunk 1]
```diff
@@ -71,6 +71,8 @@ const sortSliceTranscriptHistory = (transcriptHistory: TranscriptJSON[]): Transc
         .slice(0, SAVE_MAX_TRANSCRIPT_HISTORY)
 
 export const useChatStoreState = create<CodyChatStore>((set, get): CodyChatStore => {
+    const needsEmailVerification =
```
Comment (mrnugget) [inline client/web/src/cody/stores/chat.ts:74]: Why not use the `useIsCodyEnabled` hook here?

File: client/web/src/global/GlobalAlerts.tsx [hunk 2]
```diff
@@ -118,8 +108,8 @@ export const GlobalAlerts: React.FunctionComponent<Props> = ({ authenticatedUser
                 </DismissibleAlert>
             )}
             <Notices alertClassName={styles.alert} location="top" />
-            {!!verifyEmailProps?.emails.length && (
-                <VerifyEmailNotices alertClassName={styles.alert} {...verifyEmailProps} />
+            {isSourcegraphDotCom && (
```
Comment (mrnugget) [inline client/web/src/global/GlobalAlerts.tsx:111]: Shouldn't we only show this when the user doesn't have an email?

Also: we have a check here and then _inside_ the component too with more logic (using the hook) -- I'd rather have that in one place

File: client/web/src/global/Notices.tsx [hunk 3]
```diff
@@ -95,35 +94,24 @@ interface VerifyEmailNoticesProps {
...
-            (email): Notice => ({
-                message: `Please, <a href="${settingsURL}/emails">verify your email</a> <strong>${email
-                    .split('@')
-                    .join('\\@')}</strong>`,
-                location: 'top',
-                dismissible: false,
-            })
+    const codyEnabled = useIsCodyEnabled()
+
+    if (codyEnabled.needsEmailVerification && authenticatedUser) {
```
Comment (mrnugget) [inline client/web/src/global/Notices.tsx:101]: Shouldn't we only show this if the user does *not* have a verified email?

File: client/web/src/user/settings/emails/UserEmail.tsx [hunk 4]
```diff
@@ -28,6 +28,36 @@ interface Props {
     onEmailResendVerification?: () => void
 }
 
+export const resendVerificationEmail = async (
+    userID: string,
+    email: string,
+    options?: { onSuccess: () => void; onError: (error: ErrorLike) => void }
+): Promise<ResendVerificationEmailResult | ErrorLike> => {
```
Comment (mrnugget) [inline client/web/src/user/settings/emails/UserEmail.tsx:35]: Callbacks and Promise that returns `Result | ErrorLike` -- why both?

**Review by mrnugget — COMMENTED** [review]
I'll leave final approval to @philipp-spiess who's better with frontend than me, but I'd love to see less duplication of the boolean conditions

**Review by mrnugget — COMMENTED** [review]
I'll leave final approval to @philipp-spiess who's better with frontend than me, but I'd love to see less duplication of the boolean conditions

File: client/web/src/cody/stores/chat.ts (hunk 1 above)
Comment (thenamankumar) [reply client/web/src/cody/stores/chat.ts:74]: Can we use a hook inside the init function of zustand? Should we? @philipp-spiess?

I thought it won't be reactive.

Comment (philipp-spiess) [reply client/web/src/cody/stores/chat.ts:74]: > Can we use a hook inside the init function of zustand? 

I’m not sure actually. I assumed we can't that's why I wasn't using it :/

File: client/web/src/global/Notices.tsx (hunk 3 above)
Comment (thenamankumar) [reply client/web/src/global/Notices.tsx:101]: `needsEmailVerification` covers it.

File: client/web/src/user/settings/emails/UserEmail.tsx (hunk 4 above)
Comment (thenamankumar) [reply client/web/src/user/settings/emails/UserEmail.tsx:35]: Give more than asked?

Removed.

File: client/web/src/global/GlobalAlerts.tsx (hunk 2 above)
Comment (thenamankumar) [reply client/web/src/global/GlobalAlerts.tsx:111]: removed the check from here.

File: client/web/src/cody/stores/chat.ts (hunk 1 above)
Comment (thenamankumar) [reply client/web/src/cody/stores/chat.ts]: Yeah, not using the hooks, but I hear your concern @mrnugget. Extracted it out as util and used it. No more duplication of logic at two places.

**Pull request merged**